devices/AtomQrCode/QrCodeReader.cs

```diff
@@ -0,0 +1,843 @@
+        {
+            // need to set this upfront, otherwise we could missing data
+            _waitingForData = true;
+
+            SendCommand(
+                _decodingCommandBase,
+                0,
+                _decodingCommandBase.Length,
+                true);
+        }
+
+        private void SetScanTimeout(TimeSpan timeout)
+        {
+            // valid timeout values are: 1 - 3 - 5 - 10 - 15 - 20 seconds.
+
+            // value for 1000ms
+            byte[] suffix = new byte[] { 0x88, 0x0A, 0xFE, 0x95 };
+
+            if (timeout.TotalMilliseconds <= 1000)
+            {
+                // this is the default on the suffix
+            }
+            else if (timeout.TotalMilliseconds <= 3000)
+            {
+                suffix[_param1Index] = 0x1E;
+                suffix[_param2Index] = 0x81;
+            }
+            else if (timeout.TotalMilliseconds <= 5000)
+            {
+                suffix[_param1Index] = 0x32;
+                suffix[_param2Index] = 0x6D;
+            }
+            else if (timeout.TotalMilliseconds <= 10000)
+            {
+                suffix[_param1Index] = 0x64;
+                suffix[_param2Index] = 0x3B;
+            }
+            else if (timeout.TotalMilliseconds <= 15000)
+            {
+                suffix[_param1Index] = 0x96;
+                suffix[_param2Index] = 0x09;
+            }
+            else if (timeout.TotalMilliseconds <= 20000)
+            {
+                suffix[_param1Index] = 0xC8;
+                suffix[_param2Index] = 0xD7;
+            }
+            else
+            {
+                throw new ArgumentException();
+            }
+
+            SendCommandType07(suffix);
+
+            // store new timeout value
+            _lastTimeout = timeout;
+        }
+
+        private void SendCommandType07(
+            byte[] suffix,
+            bool sendWakeUp = false)
+        {
+            // buffer content length
+            int commandLength = 1;
+
+            // set command prefix
+            _commandBuffer[0] = _commandPrefix07;
+
+            // set command common sequence
+            // this type has different ending, so removing the last one 
+            Array.Copy(
+                _commandCommon,
+                0,
+                _commandBuffer,
+                commandLength,
+                _commandCommon.Length - 1);
+
+            // update counter
+            commandLength += _commandCommon.Length - 1;
+
+            // set command suffix
+            Array.Copy(
+                suffix,
+                0,
+                _commandBuffer,
+                commandLength,
+                suffix.Length);
+
+            // update counter
+            commandLength += suffix.Length;
+
+            SendCommand(
+                _commandBuffer,
+                0,
+                commandLength,
+                sendWakeUp);
+        }
+
+        private void SendCommandType08(
```

Comment:
quite some common code with only 2 little differences between both functions. Does it make sense to cluster them in one with something for those 2 small differences?

Reply:
The 07 it's chewing a byte at the end to perform an extra tweak... thought of that but ended up splitting it as I had to add more IFs there which ended up on more complex codel.